src/main/java/org/broadinstitute/hellbender/engine/ReferenceDataSource.java

```diff
@@ -33,39 +27,17 @@
      *
      * @param fastaFile reference fasta file
      */
-    public ReferenceDataSource( final File fastaFile ) {
-        if ( fastaFile == null )
-            throw new IllegalArgumentException("fastaFile must be non-null");
-
-        // Will throw a UserException if the .fai and/or .dict are missing
-        reference = CachingIndexedFastaSequenceFile.checkAndCreate(fastaFile);
+    public static ReferenceDataSourceFromFile of(final File fastaFile) {
+        return new ReferenceDataSourceFromFile(fastaFile);
     }
 
     /**
-     * Start an iteration over the entire reference. Not yet supported!
-     *
-     * See the BaseUtils class for guidance on how to work with bases in this format.
-     *
-     * @return iterator over all bases in this reference
+     * Initialize this data source using ReferenceBases and corresponding sequence dictionary.
      */
-    @Override
-    public Iterator<Byte> iterator() {
-        throw new UnsupportedOperationException("Iteration over entire reference not yet implemented");
+    public static ReferenceDataSourceFromReferenceBases of(final ReferenceBases bases, final SAMSequenceDictionary referenceSequenceDictionary) {
+        return new ReferenceDataSourceFromReferenceBases(bases, referenceSequenceDictionary);
     }
 
-    /**
-     * Query a specific interval on this reference, and get back an iterator over the bases spanning that interval.
-     *
-     * See the BaseUtils class for guidance on how to work with bases in this format.
-     *
-     * @param interval query interval
-     * @return iterator over the bases spanning the query interval
-     */
-    @Override
-    public Iterator<Byte> query( final SimpleInterval interval ) {
-        // TODO: need a way to iterate lazily over reference bases without necessarily loading them all into memory at once
-        return new ByteArrayIterator(queryAndPrefetch(interval).getBases());
-    }
```

Comment:
Restore `query()` here as a default implementation that just does a `return new ByteArrayIterator(queryAndPrefetch(interval).getBases());`